documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
-+   /// </summary>
-+   /// <remarks>
-+   /// On desktop, this uses GLFW.
-+   /// </remarks>
-+   public static class InputBackend
-+   {
-+       /// <summary>
-+       /// Creates an instance of the out-of-the-box input backend, which uses a native API to retrieve input backends using
-+       /// the given native handle.
-+       /// </summary>
-+       /// <param name="handle">The native handle created/sourced from the underlying native API used by this input backend.</param>
-+       /// <remarks>
-+       /// This method is implicitly called by the Windowing-Input integration. <br />
-+       /// On desktop, this uses GLFW. The handle refers to a <c>GLFWwindow*</c> if GLFW is in use.
-+       /// </remarks>
-+       public static unsafe IInputBackend Create(void* handle);
-+   }
+- This proposal incorporates/supersedes the Enhanced Input Events proposal.
+- This proposal assumes knowledge of the Windowing 3.0 proposal.
+- This proposal assumes knowledge of Silk.NET.Core 2.0.
+
+Please note that text marked **INFORMATIVE TEXT** does not form part of this proposal's requirements or proposed API, and is merely text to help the reader understand our intentions. Informative text is not well-defined, and may not reflect the actual implementation.
+
+Note that the parties in this proposal are:
+- **developer**: The developer using Silk.NET.
+- **end user**: The user using an application using Silk.NET, and by extension the user inputting data into the application via Silk.NET.
+
+Cases where the **user** word is used without the **end** prefix can be assumed to be referring to the **developer** rather than the **end user**.
+
+# Usage Examples
+```cs
+IWindowHandlesSource someWindow = null!;
+var inputContext = someWindow.CreateInput();
+inputContext.Gamepads.ThumbstickMove += @event =>
+{
+    Console.WriteLine($"Thumbstick {@event.Index} moved from {@event.OldValue} to {@event.NewValue}");
+};
+var isButtonDown = inputContext.Gamepads.Any(gamepadState => gamepadState.Buttons[JoystickButton.A]);
+```
+```cs
+IWindowHandlesSource someWindow = null!;
+var inputContext = new InputContext();
+inputContext.Backends.Add(someWindow.CreateInputBackend());
+// in future:
+// inputContext.Backends.Add(new OpenXRInputBackend(...));
+```
+
+# Reference Implementation
+
+Similar to Windowing 3.0, a reference implementation will be included in the main `Silk.NET.Input` package which uses the same API or family of APIs as Windowing's reference implementation. This will be exposed by the `InputWindowExtensions` class.
+
+```cs
+public static class InputWindowExtensions
+{
+    public static IInputBackend CreateInputBackend(this WindowHandles window);
+    public static IInputBackend CreateInputBackend(this IWindowHandlesSource window);
+    public static InputContext CreateInput(this WindowHandles window);
+    public static InputContext CreateInput(this IWindowHandlesSource window);
```

Comment:
Not too sure this is a good idea - it makes things easier but I feel like it doesn't make logical sense.

Reply:
The Working Group in the last meeting expressed distaste with the idea of a source generator (the only real alternative) being core to the Silk.NET Windowing-Input integration. As such, this proposal opts to just use something in Silk.NET.Core.

The Windowing reference implementation uses the same API/family of APIs as the Input reference implementation (specified above).

Reply:
I mean that this directly creates the context. This saves one line of code, and I feel it doesn't make too much sense that a window directly creates the full context.

Reply:
Oh right, this is for familiarity with 2.X and also just to improve readability of end user code. I don't think it does any harm, and users who don't care about multi-backend input probably won't want to look at it anyway.